### PR TITLE
Reduce more things

### PR DESCRIPTION
**Pull request by albinahlback**
- Reduce inclusions in `acb_dft.h`
- Remove test-function only `acb_vec_printd_index` from `acb_dft.h`
- Merge elliptic wrappers in `acb_modular`
- Merge randomisation in `nmod_mat`
- Reduce inclusions in `arb_fmpz_poly.h`
- Reduce inclusions in `arb_hypgeom.h`
- Remove `d_mat.h` inclusion in `fmpz_mat.h`. The three functions that requires `d_mat.h` will now require `d_mat.h` to be loaded first. (I can remove this one. I just think `d_mat.h` is quite unnused.)
- Cleanup `fmpz_mat.h`, `fmpq_mat.h` and `nmod_vec.h`
- Create a template header `src/dev/arb_mat_templates.h` to generate some boiler plate Arb matrix functions.

Two notes on the last point. First, I believe it was a function or two now changed the looping bounds from the output matrix's rows and columns to the input matrix's rows and columns. Second, in `acb_mat_conjugate` it previously aborted if output and input dimensions did not agree -- I removed this as it is neither documented, nor does it agree with similar code (I suppose this was a leftover from the transpose-conjugate or something).

### CODE REVIEW

**Comment by fredrik-johansson** [comment]
Hmm, not really a fan of the templates.

File: src/nmod_vec.h [hunk 1]
```diff
@@ -78,121 +78,104 @@ void _nmod_vec_swap(nn_ptr a, nn_ptr b, slong length)
...
-
-ulong _nmod_vec_dot(nn_srcptr vec1, nn_srcptr vec2,
-    slong len, nmod_t mod, int nlimbs);
-
-ulong _nmod_vec_dot_rev(nn_srcptr vec1, nn_srcptr vec2,
-    slong len, nmod_t mod, int nlimbs);
-
-ulong _nmod_vec_dot_ptr(nn_srcptr vec1, const nn_ptr * vec2, slong offset,
-    slong len, nmod_t mod, int nlimbs);
+#define NMOD_VEC_DOT(res, i, len, expr1, expr2, mod, nlimbs) \
```
Comment (fredrik-johansson) [inline src/nmod_vec.h:113]: Why the nonstandard level of indentation?

**Comment by albinahlback** [comment]
> Hmm, not really a fan of the templates.
> 
> Edit: I'm maybe fine with this if the template macros are defined immediately before use in `templates.c` itself. Otherwise, I'll risk having flashbacks from dealing with nonlocal code in `fq` and `lll`.
> 
> What I'd actually like to do with the `arb_mat` and `acb_mat` modules is make sure `gr_mat` has all the functionality and then just have thin wrappers, which could be autogenerated.

Hmm, I'm also fine with reverting the commit. Do I understand you correct in that I should just copypasta the things in `dev/arb_mat_templates.h` into `templates.c` to then remove said header?

File: src/nmod_vec.h (hunk 1 above)
Comment (albinahlback) [reply src/nmod_vec.h:113]: More compact format in headers. Can revert this.

**Comment by fredrik-johansson** [comment]
> > Hmm, not really a fan of the templates.
> > Edit: I'm maybe fine with this if the template macros are defined immediately before use in `templates.c` itself. Otherwise, I'll risk having flashbacks from dealing with nonlocal code in `fq` and `lll`.
> > What I'd actually like to do with the `arb_mat` and `acb_mat` modules is make sure `gr_mat` has all the functionality and then just have thin wrappers, which could be autogenerated.
> 
> Hmm, I'm also fine with reverting the commit. Do I understand you correct in that I should just copypasta the things in `dev/arb_mat_templates.h` into `templates.c` to then remove said header?

Yes.

**Pull request merged**